### ISSUE TITLE
POST success only the first iteration...? 

### ISSUE THREAD

**Issue by wasappi** [jdamcd/tempbase#2]
Hi, 
I have a similar code to send a POST request to firebase but it sends only the first time and then seems to be stuck just after the payload creation. I wonder if it's related to some unbalanced calls regarding the connection / disconnection of the client ?

**Comment by franzjacobi** [comment]
Can you share your source code, please? I can remember vaguely that I once had a similar problem.

**Comment by wasappi** [comment]
Thanks for you quick reply ! 
The first POST is correctly sent to firebase but as soon as the second one is triggered it seems to be stuck after Serial.println(bodyBuffer); The loop is then stopped for an obscure reason :/ 

`//
//   FILE:  dht11_test1.pde
// PURPOSE: DHT11 library test sketch for Arduino
//

#include <dht11.h>
#include <WiFi101.h>
#include <ArduinoJson.h>

#define INTERVAL 20 * 60 * 1000 // 20 mins

// Sensor
dht11 DHT11;
#define DHT11PIN 6
#define LED 6

// WiFi
#define SSID "SSID"
#define PASSWORD "Password"
WiFiSSLClient client;
int status = WL_IDLE_STATUS;

// Firebase
#define HOST "Host"
#define SECRET "Secret"
#define PATH "/home/bedroom.json"
char bodyBuffer[100];

void setup()
{
  Serial.begin(9600);  
  Serial.println("DHT11 TEST PROGRAM ");
  Serial.print("LIBRARY VERSION: ");
  Serial.println(DHT11LIB_VERSION);
  Serial.println();

  connectWiFi();
  syncRtc();
  Serial.println("RTC sync complete");
}

void connectWiFi() {
  status = WiFi.status();
  while (status != WL_CONNECTED) {
    Serial.println("Connecting to WiFi");
    status = WiFi.begin(SSID, PASSWORD);
  }
  Serial.println("WiFi connected");
}
 /*
 * loop() - this function will start after setup 
 * finishes and then repeat
 */
void loop()
{
  int chk = DHT11.read(DHT11PIN);

  Serial.print("Read sensor: ");
  switch (chk)
  {
    case DHTLIB_OK: 
    Serial.println("OK"); 
    break;
    case DHTLIB_ERROR_CHECKSUM: 
    Serial.println("Checksum error"); 
    break;
    case DHTLIB_ERROR_TIMEOUT: 
    Serial.println("Time out error"); 
    break;
    default: 
    Serial.println("Unknown error"); 
    break;
  }

  Serial.print("Humidity (%): ");
  Serial.println((float)DHT11.humidity, 2);

  Serial.print("Temperature (°C): ");
  Serial.println((float)DHT11.temperature, 2);

  Serial.print("Temperature (°F): ");
  Serial.println(Fahrenheit(DHT11.temperature), 2);

  Serial.print("Temperature (°K): ");
  Serial.println(Kelvin(DHT11.temperature), 2);

  Serial.print("Dew Point (°C): ");
  Serial.println(dewPoint(DHT11.temperature, DHT11.humidity));

  Serial.print("Dew PointFast (°C): ");
  Serial.println(dewPointFast(DHT11.temperature, DHT11.humidity));

  digitalWrite(LED, HIGH);

  connectWiFi();
  client.connect(HOST, 443);
  delay(1000);
  if (!client.connected()) {
    Serial.println("Error connecting to host");
  } else {
    Serial.println("Measuring");
    int temp = DHT11.temperature;
    int humidity = DHT11.humidity;
    sendData(temp, humidity);
    client.stop();
  }

  WiFi.disconnect();
  digitalWrite(LED, LOW);

  delay(INTERVAL);
}

void sendData(int temp, int humidity) {
  Serial.println("Preparing payload:");
  int size = constructJson(getEpoch(), temp, humidity);
  Serial.println(bodyBuffer);
  char pathBuffer[100];
  sprintf(pathBuffer, "POST %s?auth=%s HTTP/1.1", PATH, SECRET);

  client.println(pathBuffer);
  client.print("Host: ");
  client.println(HOST);
  client.println("Content-Type: application/json");
  client.print("Content-Length: ");
  client.println(size);
  client.println("Accept: application/json");
  client.println("Connection: close");
  client.println();
  client.println(bodyBuffer);
  
  printResponse();
}

void printResponse() {
  delay(1000);
  int readSize = client.available();
  while (readSize > 0) {
    Serial.print((char) client.read());
    readSize--;
  }
}

int constructJson(unsigned long timestamp, int temperature, int humidity) {
  StaticJsonBuffer<100> jsonBuffer;
  JsonObject& root = jsonBuffer.createObject();
  root["timestamp"] = timestamp;
  root["temperature"] = temperature;
  root["humidity"] = humidity;
  root.printTo(bodyBuffer, sizeof(bodyBuffer));
  return root.measureLength();
}

**Comment by franzjacobi** [comment]
Have you tried to reduce your interval yet? E.g. 10 sec. Nice to know what happens then.

**Comment by wasappi** [comment]
Yep, I've already tried a delay of 10seconds in the loop but with no luck, same behavior.
I wish I could add some breakpoints to debug this insanity :/

**Comment by franzjacobi** [comment]
What Arduino board are you running this code on? Either you are running out of SRAM or there is a problem to request the server. Actually I am thinking your code stuck at the sprintf and you are running out of RAM.
Additionally you could try to store strings with the F() macro for saving some extra RAM

**Comment by wasappi** [comment]
It's an MKR1000.
By tweaking some delays, I'm now confronted to a connection problem with Firebase after the first POST. But not always, it's a very erratic behavior.
So either I'm stuck around the sprintf either the arduino will refuse to connect to firebase after the first (sometimes the second) POST request. 
It's so frustrating : / But many thanks for your help!

**Comment by wasappi** [comment]
Quick update; 
If I only try a connection without performing any POST request with "www.google.com" as host, it succeeds everytime. If I replace host with my firebase url, it succeeds 2-3 times & then failed each time after

**Comment by wasappi** [comment]
Final update : I've ended creating a simple rails server hosted on heroku to post my requests, and it went smooth! No more unexplained deconnection, no more loophole, just happiness with my cheap homemade weather station :)

**Comment by jester178** [comment]
Is the POST only in the first iteration can't fix? Please HELP.
